### PR TITLE
Proposal: Add API Server Service Information to `KarmadaStatus`

### PR DESCRIPTION
**Pull request by jabellard**
**What type of PR is this?**

/kind design

**What this PR does / why we need it**:
- Details are contained in the proposal.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: docs/proposals/karmada-operator/api-server-service-status/README.md [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+// APIServerServiceStatus contains the current status of the API server service.
+type APIServerServiceStatus struct {
+	// Name represents the name of the service.
+	Name string `json:"name"`
+
+	// Port represents the client port exposed by the service.
+	Port int32 `json:"port"`
+}
```
Comment (RainbowMango) [inline docs/proposals/karmada-operator/api-server-service-status/README.md:75]: Is this structure suitable for all kinds of Services?
As you know, karmada-operator allows users to specify [Karmada APIServer ServiceType](https://github.com/karmada-io/karmada/blob/c45b817e7488149454a8c73a43dd93f03f3e1cfd/operator/pkg/apis/operator/v1alpha1/type.go#L270) which including `ClusterIP`, `NodePort`, `LoadBalancer`, and `ExternalName`. 

I'm not sure if all of these service approaches have been implemented or work as expected.
But I think it's a great chance to revisit the design about how to expose Karmada APIServer Service.

Any thoughts?
Also, cc @chaunceyjiang who might interested in this.

Comment (jabellard) [reply docs/proposals/karmada-operator/api-server-service-status/README.md:75]: @RainbowMango , you bring up a great point. As far as I can tell by looking at the code, there's only support for `ClusterIP`, `NodePort`, and `LoadBalancer` service types. I don't see support for service type `ExternalName`, which I expected as I don't think that would make sense for the API server. In addition to the service name and its type, let me break down what I think will be required for each type. 
1. `ClusterIP`: the port
2. `NodePort`: the port and node port
3. `LoadBalancer`: the port, node port, and load balancer status

As such, the status struct will look something like this:
```go
type APIServerServiceStatus struct {
	// Name represents the name of the service.
	Name string `json:"name"`

	// ServiceType represents the service type
	ServiceType corev1.ServiceType `json:"serviceType"`

	// Port represents the port exposed by the service.
	Port int32 `json:"port"`

	// NodePort represents the node port on which the service is exposed
	// when service is of type NodePort or LoadBalancer
	// +optional
	NodePort int32 `json:"nodePort"`

	// LoadBalancerStatus represents the status of a load-balancer when
	// the service is of type LoadBalancer
	// +optional
	LoadBalancerStatus corev1.LoadBalancerStatus `json:"loadBalancerStatus"`
}
```

Anything I may have missed? Would love to hear your thoughts?

Comment (jabellard) [reply docs/proposals/karmada-operator/api-server-service-status/README.md]: @RainbowMango , thanks for the constructive feedback. I applied the changes as we discussed easier during the meeting.

Comment (RainbowMango) [reply docs/proposals/karmada-operator/api-server-service-status/README.md]: Thank you. I will take a look as soon as possible but focus more on the external ETCD feature.

**Review by RainbowMango — COMMENTED** [review]
/assign

File: docs/proposals/karmada-operator/api-server-service-status/README.md (hunk 1 above)
Comment (jabellard) [reply docs/proposals/karmada-operator/api-server-service-status/README.md]: @RainbowMango , thanks again for your thought questions and constructive feedback. I made the changes we discussed during the meeting.

Comment (jabellard) [reply docs/proposals/karmada-operator/api-server-service-status/README.md]: @RainbowMango , thanks for pointing that out. Thought I arranged and squashed the commits, but forgot. Just pushed update to this PR to address that.

**Comment by RainbowMango** [comment]
Thanks @chaosi-zju for your effort. Yes, we noticed that currently, karmada-operator allows people to config all kinds of Service types, and @jabellard analyzed it in [the discussion above](https://github.com/karmada-io/karmada/pull/5599#discussion_r1773961675).
Glad to see the test report, I believe it can help us figure out the situation more clearly.

But, I can't access the doc, just sent a request for that :) :)

**Comment by jabellard** [comment]
> I conducted tests on different service types and documented for reference: [doc link](https://docs.google.com/document/d/1JUqg9Dy4mP4LhnQ6JHWMUjah6eSgZxe-srXjyXvhVq4/edit?tab=t.0#heading=h.yyo4147v9cda)

I'd love to learn more about your findings. I just sent you a request to get access to the doc.

**Comment by jabellard** [comment]
> I conducted tests on different service types and documented for reference: [doc link](https://docs.google.com/document/d/1JUqg9Dy4mP4LhnQ6JHWMUjah6eSgZxe-srXjyXvhVq4/edit?tab=t.0#heading=h.yyo4147v9cda)

@chaosi-zju , thank you for this. I'd love to learn more about your findings. I just sent you a request to get access to the doc.

File: docs/proposals/karmada-operator/api-server-service-status/README.md [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+	APIServerServiceStatus APIServerServiceStatus `json:"apiServerServiceStatus,omitempty"`
+}
+
+// APIServerServiceStatus contains the current status of the API server service.
+type APIServerServiceStatus struct {
+    // Name represents the name of the service.
+    Name string `json:"name"`
+    
+    // Port represents the port exposed by the service.
+    Port int32 `json:"port"`
```
[the remark this answers is not in the document]
Comment (jabellard) [reply docs/proposals/karmada-operator/api-server-service-status/README.md:74]: I think that's fine. We can omit it for now.

[the remark this answers is not in the document]
Comment (jabellard) [reply docs/proposals/karmada-operator/api-server-service-status/README.md]: @RainbowMango , I just pushes a change to address this. Any other comments?

File: docs/proposals/karmada-operator/api-server-service-status/README.md [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+## Proposal
+
+The proposal is to introduce a new field, `APIServerServiceStatus`, in `KarmadaStatus` to capture the name and port of the Karmada API server service. This would provide a more reliable method for referencing the service when creating additional resources, like `Ingress` objects.
+
+
+### User Stories
+
+#### Story 1
+As an operator, I want to provision an Ingress for the Karmada API server without needing to guess or infer the service name and port.
```
[the remark this answers is not in the document]
Comment (jabellard) [reply docs/proposals/karmada-operator/api-server-service-status/README.md:45]: Snippet of code for creating the ingress resource:
```golang
ingressRules := make([]networkingv1.IngressRule, len(hosts))
for i, host := range hosts {
ingressRules[i] = networkingv1.IngressRule{
	Host: host,
	IngressRuleValue: networkingv1.IngressRuleValue{
		HTTP: &networkingv1.HTTPIngressRuleValue{
			Paths: []networkingv1.HTTPIngressPath{
				{
					Path:     "/",
					PathType: ptr.To(networkingv1.PathTypePrefix),
					Backend: networkingv1.IngressBackend{
						Service: &networkingv1.IngressServiceBackend{
							Name: serviceName,
							Port: networkingv1.ServiceBackendPort{
								Number: apiServerPort,
							},
						},
					},
				},
			},
		},
	},
}
}

ingress.Annotations = map[string]string{
"nginx.ingress.kubernetes.io/backend-protocol": "HTTPS",
"nginx.ingress.kubernetes.io/ssl-passthrough":  "true",
}
ingress.Spec = networkingv1.IngressSpec{
IngressClassName: ptr.To(constants.DefaultIngressClassName),
Rules:            ingressRules,
}
```

The most relevant part is where I have to configure the backend service for the ingress rule. In there, I need the service name and port.

[the remark this answers is not in the document]
Comment (RainbowMango) [reply docs/proposals/karmada-operator/api-server-service-status/README.md:45]: Thanks. Do you think it make sense to have this in this proposal?

File: docs/proposals/karmada-operator/api-server-service-status/README.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+	// +optional
+	APIServerServiceStatus APIServerServiceStatus `json:"apiServerServiceStatus,omitempty"`
+}
+
+// APIServerServiceStatus contains the current status of the API server service.
+type APIServerServiceStatus struct {
+    // Name represents the name of the service.
+    Name string `json:"name"`
+}
+```
```
Comment (RainbowMango) [inline docs/proposals/karmada-operator/api-server-service-status/README.md:73]: Maybe we can rename the filed to `APIServerService` which is kind of a reference now, not reporting any detail status now, and add some comments, like:
```golang
// KarmadaStatus define the most recently observed status of the Karmada.
type KarmadaStatus struct {
	// APIServerService reports the location of the Karmada API server service which
	// can be used by third-party applications to discover the Karmada Service, e.g.
	// expose the service outside the cluster by Ingress.
	// +optional
	APIServerService *APIServerService `json:"apiServerService,omitempty"`
}

// APIServerService tells the location of Karmada API server service.
// Currently, it only includes the name of the service. The namespace
// of the service is the same as the namespace of the current Karmada object.
type APIServerService struct {
	// Name represents the name of the Karmada API Server service.
	// +required
	Name string `json:"name"`
}
```

**Review by RainbowMango — COMMENTED** [review]
Hi @jabellard 
Generally looks good to me.

From the test report @chaosi-zju provided on above, we can tell that:
1. Currently we supports 2 Service types, there are `ClusterIP` and `Loadbalancer`.
2. The `Loadbalancer` is not supported yet, but I think it makes sense to enhance it this time. See the use case on #4634.

So I just opened #5767 to track all of this. Any thoughts?

File: docs/proposals/karmada-operator/api-server-service-status/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (jabellard) [reply docs/proposals/karmada-operator/api-server-service-status/README.md:45]: Ok. I'll add it.

File: docs/proposals/karmada-operator/api-server-service-status/README.md (hunk 4 above)
Comment (jabellard) [reply docs/proposals/karmada-operator/api-server-service-status/README.md:73]: Ok. I'll make the change.

**Comment by jabellard** [comment]
> Hi @jabellard Generally looks good to me.
> 
> From the test report @chaosi-zju provided on above, we can tell that:
> 
> 1. Currently we supports 2 Service types, there are `ClusterIP` and `Loadbalancer`.
> 2. The `Loadbalancer` is not supported yet, but I think it makes sense to enhance it this time. See the use case on [When installing karmada using the operator, provide a way to annotate karmadaAPIServer services #4634](https://github.com/karmada-io/karmada/issues/4634).
> 
> So I just opened #5767 to track all of this. Any thoughts?

@RainbowMango , thanks for sharing this and also to @chaosi-zju  for generating this thorough report. I'll keep on eye on these action items. I just pushed some changes to address your comments.

**Comment by RainbowMango** [comment]
Thanks Joe, I think we are ready to go after [squashing the commits](https://karmada.io/docs/next/contributor/github-workflow/#squash-commits).

We can start to implemente it now.

**Comment by jabellard** [comment]
> Thanks Joe, I think we are ready to go after [squashing the commits](https://karmada.io/docs/next/contributor/github-workflow/#squash-commits).
> 
> We can start to implemente it now.

Just squashed. I started the implementation, and will send a PR for that soon.

**Review by RainbowMango — APPROVED** [review]
/lgtm
/approve

**Comment by RainbowMango** [comment]
The failing test is unrelated. We can retest it by leaving a comment `/retest` after all tests are finished.

**Comment by RainbowMango** [comment]
Right, thanks. Updated on https://github.com/karmada-io/karmada/pull/5599#pullrequestreview-2409681400.

**Pull request merged**